add a KEP for Gang Jobsts following latest developments with workload builder (#1253)
kubernetes-sigs/jobset: keps/969-WAS-integration/README.md

constraints, disruption modes, and ReplicatedJob-level resource claims. The feature is opt-in and
behind the `WorkloadAwareScheduling` alpha feature gate.

[kep-6089]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/6089-was-controller-apis

## Motivation

Distributed training and HPC workloads often need all pods admitted together, placed within a
specific topology domain, or disrupted as a group. JobSet currently has no native API for these
requirements. WAS supplies standard scheduling APIs and a `workloadbuilder` library that JobSet
can compile into `Workload` and `PodGroup` resources.

### Goals

- Add optional, centralized scheduling configuration to `JobSetSpec`.
- Support Basic and Gang policies, topology constraints, disruption modes, and shared
  ReplicatedJob-level resource claims.
- Support both one gang for the entire JobSet and independent gangs per ReplicatedJob.
- Make the scheduling configuration available to integrations such as Kueue, so they can
  select the appropriate workload representation and admission behavior.
- Preserve existing behavior when scheduling is omitted.

> 

Reviewer: One of the goal should be to support scheduling library that @mm4tt added as part of 1.37, right?

Author: yea, done.

### Non-Goals

- Implementing scheduler-side Gang, TAS, or WAP behavior.
- Implementing Kueue admission or queue management. Kueue integration is limited to exposing
  enough scheduling configuration for Kueue to understand the JobSet's workload shape; a
  follow-up Kueue design must define queueing and partial-eviction behavior.
- Adding scheduling fields to the upstream Job API.

> - Creating a composite/parent PodGroup hierarchy in the alpha implementation.

Reviewer: I think we should at least make our API extensible to support CPG. I believe, we are going to have use-cases when hierarchy might be needed of group of Jobs within JobSet.

Author: yea I'm having a hard time figuring out the API for CPG and JobSet though. 

The code just merged last week so trying to understand how we could use this in JobSet.

I still think this KEP, as is, is useful and maybe we can consider CPQ as a followup. I'm open to breaking the API for scheduling apis as we iterate. Its under an alpha feature gate so I think that is fine.

Author: A rough sketch, we are going to need to specify groups of replicated jobs that share the same podgroups. 

The other question is what percentage of our users would benefit from simple all or nothing scheduling for JobSets. For Kueue, we basically only do admission / preemption as a full gang so we could consider the happy path.

Reviewer: >yea I'm having a hard time figuring out the API for CPG and JobSet though.

@yankay do you have any initial thoughts how API design might look like for LWS and DisaggregatedSet to support Composite Pod Groups?

>The other question is what percentage of our users would benefit from simple all or nothing scheduling for JobSets. 

I don't think we need any API changes to support this.

Author: > I don't think we need any API changes to support this.

Literally this is this KEP. Maybe I misspoke. The simplest use case I want to solve is how do we specify TAS, DRA, Preemption, Gang for a jobset as a whole (not sub gangs). This seems to match the majority of use cases around standalone JobSet.

- Delegated lifecycle management of scheduling objects.

## Proposal

Add `spec.scheduling` to `JobSetSpec`. Configuration is centralized at the JobSet level and can
specify defaults for all ReplicatedJobs plus targeted overrides using
`targetReplicatedJob`. When enabled, the JobSet controller builds one `Workload`, materializes
one `PodGroup` per scheduling template, and maps child Jobs and Pods to those PodGroups.

### User Stories

| User need | Behavior |
|---|---|
| Gang-schedule a complete training JobSet | A top-level Gang policy creates one PodGroup whose `minCount` is the total pod count. |
| Schedule groups independently | A `replicatedJobPolicies` entry creates one PodGroup per targeted ReplicatedJob; Basic, Gang, topology, disruption, and resource-claim settings can be overridden. |

> | Use an independent driver and constrained workers | The driver can use Basic scheduling while workers use a topology-constrained Gang PodGroup. These groups are independent in alpha; there is no composite PodGroup. |

Reviewer: There is also use-case when MPI Launcher and Workers should share the same PodGroup, when Launcher runs as Worker. Ref Runtime:
https://github.com/kubeflow/trainer/blob/master/manifests/base/runtimes/deepspeed_distributed.yaml

Author: Its not clear to me if Trainer would leverage this integration or they would have their own representation for Gangs.

In this case, you could model this runtime as a simple gang (driver + worker are preempted together / schedulged together). That would get us pretty far.

Reviewer: We still need to add `spec.scheduling` API for TrainJob and TrainingRuntime to support it as part of: https://github.com/kubeflow/trainer/pull/3219.

>In this case, you could model this runtime as a simple gang (driver + worker are preempted together / schedulged together).

Yeah, but what if this Runtime also has initialization and post-processor ReplicatedJob? That should be separated to a different PodGroup.

| Keep existing JobSets unchanged | A JobSet without `spec.scheduling` creates no WAS resources. |

> | Avoid reserving resources for suspended workloads | Suspended JobSets have their Workload and PodGroups deleted; they are recreated on resume. |

Reviewer: @helayoty Do we have the same logic for Job? E.g. Suspended Job deletes corresponding Workload and PodGroup?

Author: It is planned for beta for Job. Right now suspend doesn't remove workload / podgroup.

Reviewer: Sounds good.

> | Combine sequencing with Gang scheduling | `DependsOn` and `InOrder` startup use one PodGroup per ReplicatedJob to avoid deadlock. |

Reviewer: Let's say we have [Initializer jobs + MPI.](https://github.com/kubeflow/trainer/blob/master/manifests/base/runtimes/torchtune/qwen2_5/qwen2_5_1.5B.yaml) That might require initializer + trainer PodGroups. Trainer PodGroup consists of two rJobs: Launcher + Node.

Author: Yea this is probably a good use case for CPG. Though you could get by with independent Rj per PodGroup and failure policy handles case if launcher dies.

| Scale an elastic workload | Changes to the represented pod count update the generated Gang `minCount`; other template changes recreate the owned scheduling resources. |
| Share a DRA claim across a Job's pods | A `resourceClaims` entry on a ReplicatedJob policy is copied to that ReplicatedJob's PodGroup, and each pod references that shared claim explicitly. Pod-level claims such as a per-pod GPU claim remain independent. |

> | Run a TPU multi-slice workload | A JobSet can use a top-level Gang and topology constraints to request coordinated placement for TPU slices. The TPU-specific pod resources and topology values remain in the JobSet pod templates. |

Reviewer: We need @GiuseppeTT @jianqiaol feedback here. IIRC, in case of JAX every replica within ReplicatedJob might consume its own PodGroup and ResourceClaim.

Author: yes, I'm open to feedback from google but I don't really want to block all implementation / merging on this.

From @GiuseppeTT's discussion, most of their users use simple replicated jobs. So even if this is a potential use case I am not sure we need to cover everything right away.

https://docs.cloud.google.com/kubernetes-engine/docs/how-to/tpu-multislice

Reviewer: Maybe @richabanker, @siyuanfoundation @fedebongio can provide additional feedback. How important is this for WAS integration?
Have you heard of any user requirements where each replica of a single ReplicatedJob needs to be assigned its own dedicated ResourceClaim? For WAS that will require to create dedicated PodGroupTemplate per replica for ReplicatedJob.

Reviewer 2: /cc @mortent , can you take a look and provide feedbacks?

Reviewer 3: @andreyvelich You recall correctly. In JAX TPU multi-slice, the `ReplicatedJob` represents a 'slice', and we use `replicas: N` to stamp out $N$ slices. 

Each replica strictly requires its own `PodGroup`. This is necessary so the topology-aware scheduler can place each replica onto a distinct, contiguous hardware mesh (e.g., a 4x4 TPU block). A single top-level gang obscures these boundaries, rendering this user story insufficient for real multi-slice deployments.

Author: kk. sounds like we are well aligned then.

| Recover one failed component | Partial eviction and rescheduling of an individual ReplicatedJob remain future work; alpha restarts follow JobSet's existing group-level semantics. |

## Risks and Mitigations

| Risk | Mitigation |
|---|---|
| Upstream `scheduling.k8s.io/v1alpha3` APIs are unstable | Keep the feature alpha and disabled by default. |
| Alpha has no composite PodGroup | Document that only top-level Gang without overrides or sequencing is JobSet-wide atomic; evaluate a hierarchy before graduation. |
| Workload and PodGroup template specs change | Patch the generated Gang `minCount` for represented pod-count changes; detect other template drift and delete/recreate the owned resources. |
| `workloadbuilder` API changes | Isolate the mapping code and pin the Kubernetes dependency version. |

## Design Details

### API

`JobSetScheduling` contains composite defaults and targeted ReplicatedJob policies:

```go
type JobSetScheduling struct {

>     Policy              *schedulingv1alpha3.PodGroupSchedulingPolicy `json:"policy,omitempty"`

Reviewer: @helayoty @mm4tt @kannon92 Maybe I missed that, but why did we add `scheduling...` prefix to the Job API?
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/batch/v1/types.go#L510-L519

For me this API is clearer:
```yaml
  scheduling:              
    policy: ...
    constraints: ...
    disruptionMode:  ...
 ```

Author: Not sure where I found it but this was requested during api-review.

    Constraints         *schedulingv1alpha3.PodGroupSchedulingConstraints `json:"constraints,omitempty"`

>     Disruption          *schedulingv1alpha3.DisruptionMode `json:"disruption,omitempty"`

Reviewer: ```suggestion
    DisruptionMode          *schedulingv1alpha3.DisruptionMode `json:"disruptionMode,omitempty"`
```

Author: done.

>     ReplicatedJobPolicies []ReplicatedJobSchedulingPolicy `json:"replicatedJobPolicies,omitempty"`

Reviewer: I think, I mentioned this before but why do need to separate replicatedJobPolicies in its own API?
This might be clearer approach:
Applied to the all rJobs similar to failurePolicies.

```yaml
scheduling:
  targetReplicatedJobs: []
  policy:
    gang: {}
```
Applied to specific rJobs

```yaml
scheduling:
  - targetReplicatedJobs: ["initializer"]
     policy:
        basic: {}
  - tagetReplicatedJobs: ["launcher", "node"]
     policy:
        gang: {}
```

The only open questions:
1. What is users want to configure PodGroup per Replica?
(Potentially, we can add `groupBy: Replica / ReplicatedJob` parameter later.
3. How to support CPG?

Author: > What is users want to configure PodGroup per Replica?
(Potentially, we can add groupBy: Replica / ReplicatedJob parameter later.

I haven't really seen anyone mention that this is a strong use case. WAS also limits the number of podgroups to 8 and I haven't really heard anyone in JobSet community say this is a strong limitation. 

If we start to have each replica be a separate PodGroup with different topology requirements we put a limit of the number of replicas to be what WAS supports.

Author: I like this API for CPG future proof.

Author: @mimowo @mwielgus 

WDYT of this API? For Kueue I could see a world where we want groups of replicated jobs to be admitted/preempted together within a JobSet.

Author: I drafted this as an alternative API for now.

Reviewer 4: My thoughts:

1. Initially, I liked this API because it seems to cleanly map to a 2-level CPG model:

```yaml
spec:
  scheduling:
    schedulingPolicy: gang {} # CPGTemplate level
     - targetReplicatedJobs: ["initializer"]
        # PGTemplate level

```

However, considering JobSet is actually a 3-level controller (JobSet -> ReplicatedJob -> Job), this mapping is incomplete. It only maps cleanly if we assume one very specific 2-level use case is valid (see point 3).

2. Regarding configuring PodGroups per replica and the 8-PodGroup limit:
To clarify, there is a strict distinction between the template hierarchy (`CPGTemplate` / `PGTemplate` in the Workload API) and the runtime hierarchy (`CPG` / `PG`). The 8-element limit applies exclusively to the template side (e.g., the number of `replicatedJobs`), not the runtime instances. You can have a `ReplicatedJob` with 100 replicas, resulting in 1 `PodGroupTemplate` and 100 `PodGroup` instances. This is fully supported in WAS. Furthermore, there is no technical value in combining multiple Jobs into a single runtime `PodGroup`; we should default to creating a new `PodGroup` per Job.
3. If you think about it - JobSet actually operates as a 3-level controller (JobSet -> ReplicatedJob -> Job).
Depending on user requirements (using topology as an example, but we could use disruptionMode or schedulingPolicy that needs to be different per level), users might require 2 or 3 configuration levels:

* **3 levels:** JobSet in a zone, ReplicatedJob in a block, Job in a rack. (Root `CPGTemplate`: `topologyKey=zone`, ReplicatedJob `CPGTemplate`: `topologyKey=block`, Job `PGTemplate`: `topologyKey=rack`).
* **2 levels (A):** JobSet in a zone, ReplicatedJob in a rack (all Jobs for that ReplicatedJob share the same rack). (Root `CPGTemplate`: `topologyKey=zone`, ReplicatedJob `CPGTemplate`: `topologyKey=rack`, Job `PGTemplate`: no topology, inherits from parrent).
* **2 levels (B):** JobSet in a zone, Job in a rack (Jobs within a ReplicatedJob can span different racks). (Root `CPGTemplate`: `topologyKey=zone`, ReplicatedJob `CPGTemplate`: no topology, Job `PGTemplate`: `topologyKey=rack`).

If we assume only "2 levels (A)" will ever be needed, the proposed API is sufficient. If all 3 are valid, the API must be extended to express configuration at all three tiers (root CPG, per-ReplicatedJob CPG, and per-Job PG).

Reviewer: > However, considering JobSet is actually a 3-level controller (JobSet -> ReplicatedJob -> Job), 

Yes, good point. @carreter proposed before to have dedicated CRD for ReplicatedJob, but we decided to not go ahead with that: https://github.com/kubernetes-sigs/jobset/issues/946

>If all 3 are valid, the API must be extended to express configuration at all three tiers

I think, all 3 might be valid in the future. As I mentioned in the previous Google documents, theoretically users can represent even disaggregated inference workloads using JobSet: https://docs.google.com/document/d/1aU1v5qxBi12YdP5cAIZXMvbrf-3tN5SIQHPwPDcKhNM/edit?disco=AAABywJxDLo 

Maybe we can think about API similar to this:
```yaml
spec:
  scheduling:
    # Level 1 — whole JobSet -> root CompositePodGroup
    policy: { gang: {} }
    constraints: { topology: [{ level: topology.kubernetes.io/zone }] }
    disruptionMode: { all: {} }

    replicatedJobPolicies:
      - targetReplicatedJobs: ["launcher", "node"] # selector, JobSet convention preserved
        # Level 2 — spans the Jobs (replicas) of each targeted rJob -> one CPG per rJob
        replicatedJob:
          policy: { gang: {} }
          constraints: { topology: [{ level: .../block }] }
        # Level 3 — one Job -> one leaf PodGroup
        job:
          policy: { gang: { minCount: 8 } }
          constraints: { topology: [{ level: .../rack }] }
          resourceClaims: [...]
```

To group multiple rJobs, I think we can [leverage our `groupName` API.](https://github.com/kubernetes-sigs/jobset/blob/main/api/jobset/v1alpha2/jobset_types.go#L327)
```yaml
apiVersion: jobset.x-k8s.io/v1alpha2
kind: JobSet
spec:
  replicatedJobs:
    # ── Stage 1: data / model download ────────────────────────────────────────
    - name: initializer
      groupName: init

    # ── Stage 2: MPI. launcher is rank 0 AND participates in the collective ───
    - name: launcher
      groupName: trainer

    - name: node
      groupName: trainer

  scheduling:
    # ── Level 1 — whole JobSet -> root CompositePodGroup ──────────────────────
    policy: { basic: {} } # stages admit independently (init completes first)
    constraints: { topology: [{ level: topology.kubernetes.io/zone }] }
    disruptionMode: { single: {} } # a finished initializer must not block/evict the trainer

    # ── Level 2 — per ReplicatedJob group ────────────────────────────────────
    replicatedJobGroupPolicies:
      - targetGroups: ["trainer"]
        podGroup:
          policy: { gang: {} } # minCount = numNodes
          constraints: { topology: [{ level: topology.k8s.io/rack }] }
          disruptionMode: { all: {} } # launcher + workers preempted/restarted together
          resourceClaims:
            - name: imex-channel # ONE generated claim, shared by all 9 pods
              resourceClaimTemplateName: imex-channel-template

```

Reviewer: Looking at TPU multi-slice, we might need something like this, where we create PodGroup per Replica:
```yaml
apiVersion: jobset.x-k8s.io/v1alpha2
kind: JobSet
metadata:
  name: multislice-job
spec:
  failurePolicy: { maxRestarts: 4 }
  replicatedJobs:
    - name: slice
      replicas: 3 # 3 slices
      template:
        spec:
          parallelism: 4 # 4 nodes per slice (4x4 topology, 4 chips/VM)
          completions: 4
          backoffLimit: 0
          template:
            spec:
              hostNetwork: true
              dnsPolicy: ClusterFirstWithHostNet
              nodeSelector:
                cloud.google.com/gke-tpu-accelerator: tpu-v5-lite-podslice
                cloud.google.com/gke-tpu-topology: 4x4
              containers:
                - name: jax-tpu
                  image: us-docker.pkg.dev/cloud-tpu-images/jax-ai-image/tpu:latest
                  resources: { limits: { google.com/tpu: 4 } }
  scheduling:
    # Level 1 — all 3 slices admitted together (synchronous multicontroller training)
    policy: { gang: {} } # minGroupCount defaults to 3
    disruptionMode: { all: {} } # matches backoffLimit:0 + whole-set restart
    replicatedJobPolicies:
      - targetReplicatedJobs: ["slice"]
        # Level 3 omitted — single rJob, so the root already spans the slices
        job: # one leaf PodGroup per slice
          policy: { gang: {} } # minCount defaults to 4
          constraints:
            topology: [{ level: cloud.google.com/gke-nodepool }] # one slice per node pool
          disruptionMode: { all: {} }

```

So we have this placing:
```
  CompositePodGroup/multislice-job          gang(minGroupCount=3), all
  ├─ PodGroup/multislice-job-slice-0        gang(minCount=4), topology=gke-nodepool
  ├─ PodGroup/multislice-job-slice-1        gang(minCount=4), topology=gke-nodepool
  └─ PodGroup/multislice-job-slice-2        gang(minCount=4), topology=gke-nodepool
```

@GiuseppeTT @jianqiaol Please let me know if my understanding of TPU multi-slice is correct here.

Author: For 1.37, I really want to target the beta APIs so I'd prefer maybe a separate feature gate for CompositePodGroup support.

For me, I have two goals with this integration: 
- start building a plan for https://github.com/kubernetes-sigs/kueue/issues/13707
- Support Gang JobSets on 1.37 cluster with only beta apis enabled.

Selfishly, there is no way I can offer support on alpha APIs so coupling this feature with compositePodGroups blocks the use of this feature until 1.38.

Author: I'm happy to iterate on a forward thinking design but I want to distance the design from requring CompositePodGroups.

Reviewer: I am fine to start just with PodGroups, but it would be better to avoid API breaking changes to support CPG. What if we follow my API approach, but not introduce any API fields to enable CPG, and only create PodGroups initially?
E.g. the API will look as follows:
While JobSet as a gang:
```yaml
scheduling:
  replicatedJobPolicies:
    - podGroup: # no targets -> all rJobs, one PodGroup
      policy: { gang: {} } # minCount = Σ parallelism × replicas
```

JobSet as MPI cluster, one PodGroup for Launcher and Node
```yaml
scheduling:
  replicatedJobPolicies:
    - targetReplicatedJobs: ["launcher", "node"]
      podGroup:
        policy: { gang: {} }
        constraints: { topology: [{ level: topology.k8s.io/rack }] }
```

TPU Multislice — per-Job PodGroups.
```yaml
scheduling:
  replicatedJobPolicies:
    - job: # no targets -> all rJobs, one PodGroup per Job
      policy: { gang: {} } # minCount = parallelism
      constraints: { topology: [{ level: cloud.google.com/gke-nodepool }] }
```

Author: I updated the API and KEP with a similar api.

Reviewer 3: @andreyvelich Confirming that your diagram accurately reflects the TPU multi-slice architecture. The workload requires exactly that layout: a top-level `CompositePodGroup` for synchronous admission, orchestrating $N$ child `PodGroups` (one per replica) to enforce topology constraints.

@kannon92 While excluding `CompositePodGroups` from this initial Alpha is a reasonable scope boundary, we should reconsider the updated KEP text proposing **'one JobSet-wide PodGroup'** to support TPU multi-slice. 

Flattening $N$ distinct slices into a single `PodGroup` hides the hardware boundary requirements from the downstream topology scheduler, preventing it from mapping the pods to separate TPU meshes. Since multi-slice fundamentally requires hierarchical scheduling, we should likely defer it as a supported use case until CPGs are fully integrated.

Author: Updated to reflect the TPU use case with separate PodGroups per group.

  claims:
    - name: imex-channel
      resourceClaimName: <shared-imex-claim>
    - name: gpu
      resourceClaimTemplateName: gpu-template
```

The shared claim is declared once on the ReplicatedJob's PodGroup policy and referenced by each
pod; the GPU claim is independently created for each pod.

### Defaulting

Defaulting applies only when `spec.scheduling` is present:

- Normal startup defaults a missing composite policy to Gang.
- Per-ReplicatedJob mode defaults an unset leaf policy to Gang.
- Top-level Gang `minCount` defaults to the sum of `parallelism × replicas` across all
  ReplicatedJobs.
- Per-ReplicatedJob Gang `minCount` defaults to `parallelism × replicas` for that
  ReplicatedJob.

For sequenced startup, the webhook leaves the composite policy unset and the builder applies
per-ReplicatedJob Gang defaults. An explicit composite Gang `minCount` is not used in that mode.

### Validation

The validating webhook and `workloadbuilder` enforce that:

- The feature gate is enabled.
- Every policy target names an existing ReplicatedJob and targets are unique.
- Policies are Basic or Gang; disruption modes are Single or All.
- `spec.scheduling` is immutable after creation.

> - Top-level and per-ReplicatedJob `minCount` values do not exceed their represented pod counts.

Reviewer: How are we going to deal with elastic JobSets? Are we planning to re-validate it at the runtime ?

Author: I guess that a user would need to update both MinCount in scheudling and job parallelism.

If they use the default setting, maybe it should work. As of now I recreate the podgroup/workload if the template changes. I did this mostly because elastic Jobs wasn't supported until a few weeks ago.

> - A single top-level PodGroup uses one `priorityClassName` across all ReplicatedJobs.

Reviewer: Why do we validate this?

Author: I think this is due to limitations in WAP. https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/5710-workload-aware-preemption#pod-group-priorities

### Test Plan

- **Unit:** builder modes, defaults, minCount calculations, overrides, resource claims,
  sequencing, naming, and immutable-resource recreation.
- **Webhook:** invalid targets, duplicate targets, policy and minCount validation, priority
  consistency, sequencing restrictions, immutability, and feature-gate behavior.
- **Integration:** Workload/PodGroup creation, ownership, Basic-only top-level mode,
  suspend/resume, scaling, DRA claim propagation, and gate-disabled behavior.
- **E2E:** top-level Basic and Gang, per-ReplicatedJob policies, suspension, sequencing,
  preemption, DRA claims, and scaling on a WAS-enabled cluster.

### Graduation Criteria

#### Alpha

- Feature gate, API, defaulting, and validation are implemented.
- The controller creates Workloads and PodGroups through `workloadbuilder`.
- Unit, integration, and WAS-enabled E2E coverage is available.

#### Beta

- Early-adopter feedback is collected.
- Gang and topology behavior is verified on a real WAS-enabled cluster.
- Composite hierarchy and delegated lifecycle management are evaluated.
- The feature gate may be enabled by default after the API and lifecycle design stabilize.

> 

Author: This is not relevant for JobSet.

#### Beta

- Early-adopter feedback is collected.
- Gang and topology behavior is verified on a real WAS-enabled cluster.
- Composite hierarchy and delegated lifecycle management are evaluated.
- The feature gate may be enabled by default after the API and lifecycle design stabilize.

## Implementation History

- 2026-06-28: KEP created and alpha implementation added, including Workload/PodGroup
  integration, per-ReplicatedJob scheduling, sequenced-startup fallback, lifecycle handling,
  immutable-spec recreation, and DRA resource-claim propagation.

## Drawbacks

- Depends on pre-stable `scheduling.k8s.io/v1alpha3` APIs.
- Adds scheduling resources and reconciliation complexity to the JobSet controller.
- Per-ReplicatedJob and sequenced modes are independently admitted in alpha because no composite
  PodGroup hierarchy is created.

## Alternatives

### Template Delegation Model

KEP-6089 also permits scheduling configuration inside an embedded Job template. This KEP rejects
that model because it splits configuration across API levels, depends on upstream Job API
changes, and does not match JobSet's existing targeted-policy pattern. Centralized configuration

> keeps the API in one place and avoids that dependency.

Reviewer 4: This makes sense and aligns with KEP-6089. However, to ensure a good UX, JobSet should validate that `spec.replicatedJobs[X].template.scheduling` is unset. Otherwise, users might configure both `spec.scheduling.targetReplicatedJobs` and `spec.replicatedJobs[X].template.scheduling`, which could lead to confusion. Should we add this to the validation section?

    schedulingPolicy: .. (gang within all replicas of the ReplicatedJob)
    schedulingConstants: .. (tas requirements for all replicas of ReplicatedJob)
    disruptionMode: .. (disruption within replicas of replicated job)
    jobPolicy:
      gang requirements at the job level
```

With the `JobSetGangOfGangs` feature gate enabled and the CompositePodGroups (CPG) feature,
the top level scheduling field corresponds to the root level.

With compositePodGroups, the top level policy would be a CPG object.
The group of replicated jobs within a `replicatedJobPolicy` would correspond to a CPG also.

And then at the job level, these would be separate PodGroups.

For a jobset, with three replicated jobs with 2 replicas each, gang of gangs could be represented:

JobSet (CPG, basic, zone-level topology)
  ReplicatedJobA (CPGA, preempt all) ReplicatedJobB(CPGB, preempt all) ReplicatedJobC (CPGB, preempt all)
    Job1 (PG1, gang, rack, DRA)         Job3 (PG3, gang, rack)                  Job5 (PG5, gang, rack)
    Job2 (PG2, gang, rack, DRA)         Job4 (PG4, gang, rack)                  Job6 (PG6, gang, rack)

```yaml

> scheduling:

Reviewer 3: I know the Alpha 1 implementation of this KEP focuses on generating flat `PodGroups`, but looking at the theoretical design laid out here for `JobSetGangOfGangs` in Alpha 2, I have a concern about how that future state maps to downstream schedulers.

When advanced, topology-aware scheduling plugins evaluate a `CompositePodGroup` (CPG) to map workloads across highly coupled hardware, they often rely heavily on a metric of **homogeneity**. To calculate placements efficiently without hitting a combinatorial explosion, these schedulers typically assume that all child `PodGroups` within a CPG are structurally identical (i.e., if PodGroup A fits into Hardware Slice X, then PodGroup B from the same CPG can also fit into that exact same type of hardware slice).

However, JobSet's proposed API allows for highly *heterogeneous* groupings under a single JobSet—for instance, bundling a standard CPU "driver" alongside topology-constrained accelerator "workers". If these are eventually compiled into a single CPG under the Alpha 2 design, it breaks the homogeneity assumption and forces the downstream scheduler to untangle the mixed CPU/accelerator permutations.

**Suggestion:** 
Since we are outlining the `JobSetGangOfGangs` API here, should we add a note about validating and preventing heterogeneous ReplicatedJobs from merging into a single CPG? Or, at the very least, explicitly note that downstream schedulers consuming this future API must support heterogeneous CPGs?

Author: I would check the KEP again especially the API. We can represent all the use cases you mention with the API we have. 

We can have pod groups for Jobset. Pod groups for each replicated job and pod groups for each replica of a replicated job.

This is handled in phase 1.

Author: The goal would be that eventually if you specifiy all levels (level 1, 2 and 3), each would end up being a CPG.

ie if level 1 is specified it would be both a CPG and a PodGroup probably.

But right now its hard to get all these designs in place because I find Kueue and other projects are completely ignoring WAS in integrations because the APIs are not stable. So why I focus on phase 1 is to get an actual integration that people could use in cloud environments versus discussing a million hypothetical usecases that we don't yet support in JobSet or Kueue. 

If you suggest a use case for what you have in mind, we can flesh out how that would map to CPG if you are interested in taking that on.

Reviewer: Yeah, I agree with Kevin. With the existing API design users can define PodGroups for rJobs as they want, homogeneous or heterogeneous.
I don't think we should expect that all JobSet users are rely on Kueue for scheduling and be limited to what Kueue supports today.

Reviewer 3: Thanks @kannon92 @andreyvelich. I agree with keeping Phase 1 lean and not over-constraining the API before `CompositePodGroups` reach beta upstream.

Since CPG creation is deferred to Phase 2 (Alpha 2 in 1.38), we don't need any validation or blocking changes for Phase 1. Can we just add a brief informational note under the Gang of Gangs / Future Work section acknowledging that downstream topology plugins evaluating CPG trees may require homogeneous child groups? That way, we preserve the context for when we tackle Phase 2.

Author: We do need some validation as we can't have level 1 and leve2 specified together.

So top level scheduling should be rejected if someone specifies replicatedJob scheduling also.

Reviewer 4: > When advanced, topology-aware scheduling plugins evaluate a CompositePodGroup (CPG) to map workloads across highly coupled hardware, they often rely heavily on a metric of homogeneity. To calculate placements efficiently without hitting a combinatorial explosion, these schedulers typically assume that all child PodGroups within a CPG are structurally identical (i.e., if PodGroup A fits into Hardware Slice X, then PodGroup B from the same CPG can also fit into that exact same type of hardware slice).

As an author of the CPG API, I don't think this is right. Please see my other comment - https://github.com/kubernetes-sigs/jobset/pull/1253/changes#r3830582880

| Delete | OwnerReferences provide cleanup. |
| No scheduling configuration | Create no WAS resources. |
| Gate disabled with scheduling configured | Reject the JobSet at admission. |

### Test Plan

- **Unit:** builder modes, defaults, minCount calculations, overrides, resource claims,
  sequencing, naming, and immutable-resource recreation.
- **Webhook:** invalid targets, duplicate targets, policy and minCount validation, priority
  consistency, sequencing restrictions, immutability, and feature-gate behavior.
- **Integration:** Workload/PodGroup creation, ownership, Basic-only top-level mode,
  suspend/resume, scaling, DRA claim propagation, and gate-disabled behavior.
- **E2E:** top-level Basic and Gang, per-ReplicatedJob policies, suspension, sequencing,
  preemption, DRA claims, and scaling on a WAS-enabled cluster.

### Kueue Integration

Kueue integration is tracked separately in [issue-13707](https://github.com/kubernetes-sigs/kueue/issues/13707).
Kueue has two options for queueing jobsets:

a. JobSet is admitted as a single gang
b. ReplicatedJobs are admitted separately via the Job integration

> Solving the top level JobSet is the simplest option as this basically means that all the pods within the JobSets are scheduled together.

Reviewer 3: I want to touch on the integration plan here: *"Solving the top level JobSet is the simplest option as this basically means that all the pods within the JobSets are scheduled together."*

Even with the Alpha 1 approach of generating flat `PodGroups`, I want to advocate for keeping the door open to fully exposing the underlying shape of the JobSet to Kueue in the future, rather than abstracting it away. Here is why we prefer granular visibility over a flattened representation:

As Kueue continues to mature, its preemption logic is becoming incredibly sophisticated. To minimize "wasted" disruption, Kueue needs to simulate the exact placement algorithms of the underlying schedulers (achieving bit-identical placement logic) to find the absolute *minimal set* of lower-priority pods to preempt.

If Kueue only receives a "flattened" approximation of a complex JobSet, it is blinded to the actual topological shapes and constraints of the distinct ReplicatedJobs/PodGroups inside. Without that granular footprint data, Kueue cannot accurately simulate the underlying scheduler's math, leading to sub-optimal decisions (e.g., over-preempting workloads unnecessarily, or failing to realize a preemption would result in a viable placement).

**Suggestion:**
Could we add a note to this section acknowledging that while treating the JobSet as a single top-level gang simplifies v1 admission, future iterations of Kueue integration should aim to read the full, multi-level `scheduling` specs (including the distinct `PodGroups` generated in per-ReplicatedJob mode)? This ensures Kueue preserves the data needed for high-fidelity, shape-aware preemption simulations.

Author: Yes without this APi we don't have an integration with Kueue, I have a ticket in Kueue to discuss how we should consume these workload APIs in the job framework of Kueue.

Imo Kueue has to react to this API. The Kueue integration for workloads will be a major focus once 1.37 is out but we need the APIs merged in the workload APIs to continue this discussion.

I don't think anyone in Kueue has really thought through the integration of CPG with Kueue's workload api.

Author: TBH I'm having a hard time figuring out where the put the Kueue integration. I don't think Kueue really supports partial admissions for JobSets right now so the JobSet is admitted as a single gang and preempted as a gang. The alpha 1 podgroup would solve this use case pretty well.

There are examples where customers use the Job integration within a JobSet but I'm not clear if this is really supported or recommended. I am tracking the kueue integration work in https://github.com/kubernetes-sigs/kueue/issues/13707 once we get phase 1 of this merged in a release.

My hope would be that we could enforce defaults (gangs / disruption) based on normal behavior for Kueue and figure out the rest of the story once we have this API present.

Reviewer 3: Thanks for clarifying, we definitely shouldn't block Phase 1 of this KEP on Kueue's multi-level scheduling roadmap. 

To keep the KEP accurate and set clear expectations for future work, could we just update Section 4.8 to frame this in two phases?
1. **Phase 1**: JobSet exposes the WAS scheduling configuration; Kueue integrates via standard all-or-nothing aggregate admission.
2. **Phase 2 (Tracked in kueue#13707)**: Kueue explores ingesting the multi-level `replicatedJobPolicies` to perform shape-aware preemption simulations.

| Run a TPU multi-slice workload | A JobSet can use a PodGroup per replica and topology constraints to request coordinated placement for TPU slices. The TPU-specific pod resources and topology values remain in the JobSet pod templates. |
| Recover one failed component | Partial eviction and rescheduling of an individual ReplicatedJob remain future work; alpha restarts follow JobSet's existing group-level semantics. |

### Representative APIs

Each user story above maps to a concrete `spec.scheduling` configuration and a specific set of
generated `Workload`/`PodGroup` resources.

**Gang-schedule a complete training JobSet**

```yaml
scheduling:
  schedulingPolicy:
    gang: {}
```

Creates: one `Workload` and one `PodGroup` named `<jobset-name>`, with `minCount` defaulted to
`parallelism × replicas` summed across all ReplicatedJobs.

**Schedule groups independently**

```yaml
scheduling:
  replicatedJobPolicies:
    - targetReplicatedJob: launcher
      schedulingPolicy:
        gang: {}
    - targetReplicatedJob: worker

>       schedulingPolicy:

Reviewer: @kannon92 Could you fix the API since this should be list:

```suggestion
    - targetReplicatedJobs: ["launcher"]
      schedulingPolicy:
        gang: {}
    - targetReplicatedJobs: ["worker"]
      schedulingPolicy:
```

Author: done.

MinCount for gang of launcher and worker is the number of replicas for each summed.

**Independent driver and constrained workers**

```yaml
scheduling:
  replicatedJobPolicies:
    - targetReplicatedJob: driver
      schedulingPolicy:
        basic: {}
    - targetReplicatedJob: worker
      schedulingPolicy:
        gang: {}
      schedulingConstraints:
        topologyRequest: ...
```

Creates: one `Workload` with two independent `PodGroup`s — a Basic `PodGroup` for `driver` and a
topology-constrained Gang `PodGroup` for `worker`. No composite `PodGroup` links them in alpha.

**Keep existing JobSets unchanged**

```yaml
# spec.scheduling omitted
```

Creates: no `Workload` and no `PodGroup`; child Jobs and Pods are unaffected.

**Avoid reserving resources for suspended workloads**

```yaml
spec:
  suspend: true
  scheduling:
    ...
```

Creates: the existing `Workload` and `PodGroup`(s) are deleted on suspend and recreated with the
same names/mapping on resume.

**Combine sequencing with Gang scheduling**

```yaml
scheduling:
  schedulingPolicy:
    gang: {}
replicatedJobs:
  - name: leader
    dependsOn: []
  - name: worker
    dependsOn:
      - name: leader

>         status: Ready

Reviewer: @mm4tt @kannon92 Would it be helpful to add user story that we discuss when JobSet has Initializer + Launcher + Worker jobs?

e.g.
```yaml
scheduling:
  policy: { basic: {} }
  replicatedJobPolicies:
    - targetReplicatedJobs: ["leader", "worker"]
      policy: { gang: {} }
      jobPolicy:
        policy: { gang: {} }
    - targetReplicatedJobs: ["initializer"]
      jobPolicy:
        policy: { basic: {} }

```

Author: I think this would violate our API for phase 1. For phase 2 this could be allowed.

Basically we can only specify level 1 or level 2 for phase 1. `scheduling.policy.basic` would in turn create a global PodGroup that has no gang and then we could create nested gangs for each ReplicatedJob.

Without CPG this is misleading and not allowed.

Reviewer: Make sense, let's add this use-story in the Phase 2.

Author: I'll add this when we work on Phase 2. My hope is to merge this KEp and the implementation. And release jobset 0.13. And then we can iterate on Phase 2 with alpha apis.

```

Creates: one `Workload` containing one Gang `PodGroup` per ReplicatedJob (`<jobset-name>-leader`,
`<jobset-name>-worker`) instead of a single JobSet-wide `PodGroup`, avoiding deadlock from
sequential Job creation.

**Scale an elastic workload**

```yaml
scheduling:
  schedulingPolicy:
    gang: {}
```

Creates: the existing `Workload`/`PodGroup`(s) are preserved; only the generated Gang `minCount`
is patched in place as the represented pod count changes.

**Share a DRA claim across a Job's pods**

```yaml
scheduling:
  replicatedJobPolicies:
    - targetReplicatedJob: worker
      schedulingPolicy:
        gang: {}
      resourceClaims:
        - name: imex-channel
          resourceClaimTemplateName: imex-channel-template
```

Creates: one `Workload`/`PodGroup` for `worker` whose `PodGroup` carries the `resourceClaims`
entry; each pod in `worker` references the shared claim explicitly, while per-pod claims (for
example a GPU claim) remain independent.

**Run a TPU multi-slice workload**

worker has three replicas.

```yaml
scheduling:
  replicatedJobPolicies:
  - targetReplicatedJob: worker

>     jobPolicies:

Reviewer: Should this be `jobPolicies` or `jobPolicy` since it is not the slice?

cc @mm4tt 
```suggestion
    jobPolicy:
```

Author: jobSchedulingPolicy accordinger to our API.

Reviewer: @kannon92 Any thoughts to rename `replicatedJobPolicies` to `replicatedJobSchedulingPolicies` for consistency?

Reviewer: Alternatively, we can remove policy all together to simplify API naming:
```yaml
scheduling:
  replicatedJobs:
    - targetReplicatedJobs: [..]
      job:
        schedulingPolicy: ...
```

Any thoughts @kubernetes-sigs/wg-workload-aware-scheduling-leads ?

Reviewer: cc @kannon92 @jianqiaol  on the API naming discussion.

To keep API simple I would prefer this:
```yaml
scheduling:
  replicatedJobs:
    - targetReplicatedJobs: [...]
      job:
        schedulingPolicy: {}
```
Instead of this:

```yaml
scheduling:
  replicatedJobsSchedulingPolicies:
    - targetReplicatedJobs: [...]
      jobSchedulingPolicy:
        schedulingPolicy: {}
```

Author: cc @helayoty @mm4tt @GiuseppeTT 

I'd like an agreement on the name before I change it. Honestly I'm fine with either choice. The top one is shorter.. :shrug:

Author: AFAIK this is the last item to address.

Reviewer: That's true @kannon92.
Shall we change the API design to Option 1?
```
scheduling:
  replicatedJobs:
    - targetReplicatedJobs: [...]
      job:
        schedulingPolicy: {}
```

Author: Okay. Let's just do that.

One podgroup will be created for the entire replicated job and their replicas.
This is useful when you want to control disruption or gang scheduling at the replicated job level.

For MVP, the top layer and replicatedJobs level are mutally exclusive.

For the last level, the goal would be that each replica of a ReplicatedJob should be a separate gang.
This enables the use of preemption at the replica level and provides a way for individual jobs within a JobSet
to use resourceClaims.

### Gang of Gangs Support

At a high level, the API looks like:

```yaml
scheduling:
  schedulingPolicy: .. (if gang, all pods scheduled together)
  schedulingConstants: ... (topology for all jobs within JobSet)
  disruptionMode: ... (disruption for all pods within jobset)
  replicatedJobPolicy:
    targetReplicatedJobs ..(the below policies apply to these group of replicated jobs)
    schedulingPolicy: .. (gang within all replicas of the ReplicatedJob)
    schedulingConstants: .. (tas requirements for all replicas of ReplicatedJob)
    disruptionMode: .. (disruption within replicas of replicated job)
    jobPolicy:

>       gang requirements at the job level

Reviewer: Could you fix the YAML formatting here please?

Author: I dont really know what is wrong with the formatting?

Reviewer: I think, `:` is missing after `targetReplicatedJobs: [...]`

Author: good call out. updated.

  replicatedJobPolicy:
    - targetReplicatedJobs: [ReplicatedJobA]
      disruptionMode: all
      job:
        schedulingPolicy: {}
        schedulingConstraints: rack
        DRA: ... #shared claim for all pods within podgroup

    - targetReplicatedJobs: [ReplicatedJobB]
      disruptionMode: all
      disruptionMode: all
      job:
        schedulingPolicy: {}
        schedulingConstraints: rack

    - targetReplicatedJobs: [RelicatedJobC]
      disruptionMode: all
      job:
        schedulingPolicy: {}
        schedulingConstraints: rack
        DRA: ... #shared claim for all pods within podgroup

```

### API

The api is defined below. 

```go
type JobSetScheduling struct {
    SchedulingPolicy              *schedulingv1alpha3.PodGroupSchedulingPolicy `json:"schedulingPolicy,omitempty"`
    SchedulingConstraints         *schedulingv1alpha3.PodGroupSchedulingConstraints `json:"schedulingConstraints,omitempty"`
    DisruptionMode                *schedulingv1alpha3.DisruptionMode `json:"disruptionMode,omitempty"`
    ReplicatedJobPolicies []ReplicatedJobSchedulingPolicy `json:"replicatedJobPolicies,omitempty"`
}

type ReplicatedJobSchedulingPolicy struct {

>     TargetReplicatedJob []string `json:"targetReplicatedJob,omitempty"`

Reviewer: @kannon92 @mm4tt @jianqiaol Do we need to introduce name for the policy here?
Without it, how are we going to set the name for PodGroupTemplate which represent >1 ReplicatedJob:
```yaml
targetReplicatedJobs: ["launcher", "worker"]
```
```suggestion
type ReplicatedJobSchedulingPolicy struct {
    +required
    Name string `json:"name"`
    TargetReplicatedJob []string `json:"targetReplicatedJob,omitempty"`
```

Author: For phase 1, It would mean that launcher, worker get combined into a PodGroup.

Reviewer: Yeah, but what would be the name for that PodGroupTemplate or CPGTemplate? I think, we should introduce `name` API for our policies, similar to DS API: https://github.com/kubernetes-sigs/lws/issues/894#issuecomment-5244163695

Author: In the implementation I just combine the names into a single podgroup.

So TargetReplicatedJob: [launcher, worker] with a jobset name of jobsetA will give the following name for podgroup:
jobsetA-launcher-worker.

The podGroup within the workload would be named launcher-worker.

Reviewer: @kannon92 What if I have 15 ReplicatedJobs that I want to combine into a single PodGroup?

Any particular reason you don't want to introduce `name` into ReplicatedJob policies like we did for [FailurePolicy rules](https://github.com/kubernetes-sigs/jobset/blob/main/api/jobset/v1alpha2/jobset_types.go#L444-L447)?

Author: The API is already quite complicated. We don't allow folks to set names for podGroups or Workloads for Job so I don't see a reason to give that ability to JobSet.

Reviewer: Yeah, but the APIs for Job is very simple since we only support single PodGroup for the entire workload, but JobSet APIs allow users to configure targets for PodGroups. How would you set name of PodGroup when users set many target ReplicatedJobs ?

Reviewer: As we discussed on the call, we can extend the API with the `name` later to keep API simpler in the Phase 1.

    SchedulingConstraints          *schedulingv1alpha3.PodGroupSchedulingConstraints `json:"schedulingConstraints,omitempty"`
    DisruptionMode                 *schedulingv1alpha3.DisruptionMode `json:"disruptionMode,omitempty"`
    ResourceClaims                 []schedulingv1alpha3.PodGroupResourceClaim `json:"resourceClaims,omitempty"`
    JobSchedulingPolicy            *JobSchedulingPolicy                        `json:"jobSchedulingPolicy,omitempty"`
}

type JobSchedulingPolicy struct {
    SchedulingPolicy               *schedulingv1alpha3.PodGroupSchedulingPolicy `json:"schedulingPolicy,omitempty"`
    SchedulingConstraints          *schedulingv1alpha3.PodGroupSchedulingConstraints `json:"schedulingConstraints,omitempty"`
    DisruptionMode                 *schedulingv1alpha3.DisruptionMode `json:"disruptionMode,omitempty"`
    ResourceClaims                 []schedulingv1alpha3.PodGroupResourceClaim `json:"resourceClaims,omitempty"`
}
```

`JobSetSpec.Scheduling` is optional and immutable.

### Defaulting

Defaulting applies only when `spec.scheduling` is present:

> - Normal startup defaults a missing composite policy to Gang.

Reviewer: @helayoty @kannon92 @mm4tt @tenzen-y  Do we have the same defaults for Job? Maybe we should make
`.spec.scheduling. schedulingPolicy` is required for all JobSets when `.spec.scheduling` is set?

Author: For Phase 1, we can't make any requirements. Without compositePodGroups, we actually would forbid setting `.spec.scheduling.SchedulingPolicy` and the level 2 api.

Author: Thanks for bringing this up. In my impl PR I realize we forgot the validation logic so I added that.

- Every policy target names an existing ReplicatedJob and targets are unique.
- Policies are Basic or Gang; disruption modes are Single or All.
- `spec.scheduling` is immutable after creation.
- Top-level and per-ReplicatedJob `minCount` values do not exceed their represented pod counts.
- A single top-level PodGroup uses one `priorityClassName` across all ReplicatedJobs.
- An explicit top-level Gang `minCount` is rejected with `DependsOn` or `InOrder` startup.
- An explicit Gang `minCount` cannot exceed the represented pod count, including after a
  requested downscale.
- If scheduling api is specified then there can be no scheduling api on the job api itself.
- When JobSetGangOfGangs is disabled, JobSets can not have the top level api
  and the replicatedJob api specified.
- Per-ReplicatedJob mode produces one PodGroup per resulting template. The upstream Workload
  API caps `podGroupTemplates` at 8 per Workload, while `JobSetSpec.ReplicatedJobs` has no such
  cap; the webhook rejects a JobSet in per-ReplicatedJob mode whose resulting PodGroup count
  would exceed 8, rather than partitioning across multiple Workloads or merging PodGroups.

> ### Controller Integration

Reviewer: @kannon92 Can we have section in this proposal which represent naming for Workload, PodGroupTemplate, and PodGroups?

Author: Updated with a section.

   controller owner of the Workload and PodGroups.
6. Child Jobs receive:
   - `scheduling.k8s.io/group-template-name`, identifying the template; and
   - a pod-template `schedulingGroup.podGroupName` pointing to the PodGroup.

The alpha implementation does not create a `CompositePodGroup`. It therefore does not set
`scheduling.k8s.io/parent-composite-podgroup`; that annotation refers to a
`CompositePodGroup`, not to the owning JobSet.

#### Naming Convention

The Workload is always named after the JobSet (`<jobset-name>`). Because the current
`workloadbuilder` only supports single-item builder trees, every WorkloadItem the JobSet
produces is built independently and then merged into that one Workload.

Each WorkloadItem — which compiles to one PodGroupTemplate — is named according to how it was
produced:

- **Top-level mode** (composite Gang policy, no sequenced startup, no targeted overrides): the
  JobSet name, `<jobset-name>`.
- **Per-ReplicatedJob mode** (a ReplicatedJob with no `replicatedJobPolicies` entry targeting
  it): the ReplicatedJob's own name, `<replicatedjob-name>`.
- **Grouped ReplicatedJobs** (a `replicatedJobPolicies` entry's `targetReplicatedJob` names more

>   than one ReplicatedJob): the target names joined with `-`, e.g. `<target-1>-<target-2>`.

Reviewer: That might not be possible since we don't have max number of ReplicatedJobs in the list, but PodGroup name is 63 char max, right?

Author: Yea I think in Job we may hash the name in case it goes over the limit.

<!-- /toc -->

## Summary

This KEP integrates [KEP-6089 Workload-Aware Scheduling (WAS) Controller APIs][kep-6089] into
JobSet. It adds an optional `spec.scheduling` field that lets users express gang scheduling,
topology constraints, and disruption policies for their JobSets. The feature is fully opt-in:
existing JobSets without the field are unaffected.

[kep-6089]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/6089-was-controller-apis

## Motivation

JobSet orchestrates groups of Jobs for distributed ML training and HPC workloads. These workloads
frequently need:

- **Gang scheduling:** All pods across all Jobs must be admitted together, or none at all.
- **Topology awareness:** Co-locate worker pods on the same rack or zone for network performance.
- **Coordinated disruption:** Preempting one pod should preempt the entire group.

Today, users cannot express these intents natively in a JobSet. KEP-6089 provides standardized
building blocks under `scheduling.k8s.io` and a `workloadbuilder` library. This KEP adopts those
building blocks to give JobSet users a first-class scheduling API.

> ### Goals

Reviewer 5: How does the KEP impact the integration with Kueue? Could you summarize and put the integration as Goal/non-Goal for the KEP?

Author: I think Kueue will need to read the scheduling field to figure out the kind of Kueue worklaod it needs to create.

Author: I don't know if this should be in scope for JobSet.

The goal is to make sure that JobSets can manage podgroupds/workloads in their own controller. Kueue will need separate work to read this once this merges.

I can maybe add a section about a future KEP in Kueue is needed. Basically I expect every integration with Kueue will need to follow something similar.

Author: Added a section for Kueue.

- Add an optional `spec.scheduling` field to `JobSetSpec` for expressing WAS intent.
- Support gang scheduling, topology constraints, and disruption modes at both the
  composite (whole JobSet) and leaf (per-ReplicatedJob) levels.
- Compile user intent into `Workload`/`CompositePodGroup`/`PodGroup` objects via the
  `workloadbuilder` library.
- Preserve 100% backward compatibility: JobSets without `spec.scheduling` behave identically
  to today.

### Non-Goals

> - Modifying child Job templates to inject scheduling fields (we use the centralized model).

Author: note: there was some discussion about potentially supporting this.

- Implementing the scheduler-side WAS features (Gang, TAS, WAP) — those are upstream.
- Supporting delegated `PodGroup` lifecycle management in the alpha phase.

## Proposal

Add a new optional `spec.scheduling` field to `JobSetSpec` behind a `WorkloadAwareScheduling`
feature gate. The field uses the centralized "Targeted Policies" model (Option B from KEP-6089),
where all scheduling configuration lives at the root level and targets ReplicatedJobs by name.
This aligns with JobSet's existing `targetReplicatedJobs` pattern used in `FailurePolicy` and
`SuccessPolicy`.

When `spec.scheduling` is set, the JobSet controller compiles a `Workload` resource (with a
`CompositePodGroup` root and one `PodGroup` per ReplicatedJob) and manages its lifecycle. When
`spec.scheduling` is omitted, no scheduling objects are created.

> ### User Stories

Reviewer 5: Should we add a User Story for Kueue about partial eviction (for example stop and restart only one failed Job rather than the entire JobSet)?  This is an important gap for the current integration reported by  users using large JobSets.

I'm curious how this could be realized with this integration, or this is completely orthogonal?

Author: Maybe eventually. I heard that rescheduling is going to a major push for 1.38. 
@wojtek-t @Argh4k @mm4tt

Author: I think it depends on what you mean by partial eviction. I do think I need to consider https://github.com/kubernetes/enhancements/blob/master/keps/sig-scheduling/6012-composite-podgroup-api/README.md in this design as this would allow for more granular control over preemption and sub gangs.

I haven't really thought through this integration though.

Author: PTAL at this. We are going to have two phases for this feature.

Phase 1 will be to just use Workload and PodGroup and represente podgroups within each level.

Phase 2 will be support for CompositePodGroups. CPG is alpha in 1.37 so I want to iterate on this once phase 1 is merged.

> #### Story 1: Gang-Schedule an Entire Training JobSet

Reviewer: TBD: Add use-cases for TPU multi-slice
Ref:
- https://trainer.kubeflow.org/en/latest/user-guides/jax-tpu.html#
- https://jobset.sigs.k8s.io/docs/tasks/tpu_multislice/

Author: done.

`<jobset-name>-worker`) instead of a single JobSet-wide `PodGroup`, avoiding deadlock from
sequential Job creation.

**Scale an elastic workload**

```yaml
scheduling:
  schedulingPolicy:
    gang: {}
```

Creates: the existing `Workload`/`PodGroup`(s) are preserved; only the generated Gang `minCount`
is patched in place as the represented pod count changes.

**Share a DRA claim across a Job's pods**

```yaml
scheduling:
  replicatedJobPolicies:
    - targetReplicatedJob: [worker]
      schedulingPolicy:
        gang: {}
      resourceClaims:
        - name: imex-channel
          resourceClaimTemplateName: imex-channel-template
```

Creates: one `Workload`/`PodGroup` for `worker` whose `PodGroup` carries the `resourceClaims`
entry; each pod in `worker` references the shared claim explicitly, while per-pod claims (for
example a GPU claim) remain independent.

**Run a TPU multi-slice workload**

worker has three replicas.

```yaml
scheduling:
  replicatedJobPolicies:
  - targetReplicatedJob: [worker]
    jobPolicies:
      schedulingConstraints:
        topologyRequest: ...
      gang: {}
      disruptionMode: 
        all: {}
      resourceClaim:
        - name: tpu-slice

>           resourceClaimTemplate: tpu-slice-template

Reviewer 3: The TPU multi-slice YAML example has several field and schema mismatches with the proposed Go API:
1. `jobPolicies:` should be `jobSchedulingPolicy:` to match the struct definition on line 458.
2. `gang: {}` must be nested under `schedulingPolicy: { gang: {} }`.
3. `resourceClaim:` should be plural (`resourceClaims:`).
4. `resourceClaimTemplate:` should be `resourceClaimTemplateName:`.

Author: done.

This allows one to use v1beta1 apis only for 1.37 so users can use this API sooner.

b. The use of compositePodGroups will handle under a separate feature gate as this depends on an alpha API in 1.37.

### MultiLevel support

JobSet has three layers: JobSet, ReplicatedJobs, Jobs.

Its useful to explain what each level of scheduling can buy you.

For the top layer, the main goal is to schedule a JobSet as a gang.
All the replicated jobs and their replicas are scheduled as one single podgroup.

For the ReplicatedJobs level, the goal is to control the podgroups for each ReplicatedJob.
One podgroup will be created for the entire replicated job and their replicas.
This is useful when you want to control disruption or gang scheduling at the replicated job level.

For MVP, the top layer and replicatedJobs level are mutally exclusive.

For the last level, the goal would be that each replica of a ReplicatedJob should be a separate gang.
This enables the use of preemption at the replica level and provides a way for individual jobs within a JobSet
to use resourceClaims.

### Gang of Gangs Support

At a high level, the API looks like:

```yaml

> scheduling:

Reviewer 3: There are a few YAML syntax errors, typos, and incomplete lines in this section:
1. Lines 302 & 307: `schedulingConstants` -> should be `schedulingConstraints`.
2. Line 347: typo `RelicatedJobC` -> should be `ReplicatedJobC`.
3. Lines 342–343: duplicate `disruptionMode: all` lines.
4. Inconsistent field naming between `targetReplicatedJob` and `targetReplicatedJobs`.

Author: updated.

```yaml
scheduling:
  schedulingPolicy: basic
  schedulingConstraints: zone
  disruptionMode: simple
  replicatedJobPolicies:
    - targetReplicatedJobs: [ReplicatedJobA]
      disruptionMode:
        all: {}
      jobSchedulingPolicy:
        schedulingPolicy: {}
        schedulingConstraints: rack
        resourceClaims: ... #shared claim for all pods within podgroup
    - targetReplicatedJobs: [ReplicatedJobB]
      disruptionMode: all
      disruptionMode: all
      jobSchedulingPolicy:
        schedulingPolicy: {}
        schedulingConstraints: rack
    - targetReplicatedJobs: [RelicatedJobC]
      disruptionMode: all
      jobSchedulingPolicy:
        schedulingPolicy: {}
        schedulingConstraints: rack
        resourceClaims: ... #shared claim for all pods within podgroup
```

### API

The api is defined below, matching the alpha implementation.

```go
type JobSetSpec struct {
    ...
    // scheduling defines the Workload-Aware Scheduling configuration for this JobSet.
    // When nil, no scheduling objects are created and behavior is unchanged.
    // Requires the JobSetWorkloadAwareSchedulingAPI feature gate.

>     // +kubebuilder:validation:XValidation:rule="self == oldSelf",message="Value is immutable"

Reviewer 3: There is a logical contradiction between the CEL immutability rule on line 365 and the downscale behavior described on lines 607–610:
- here we specify `+kubebuilder:validation:XValidation:rule="self == oldSelf"` (making `spec.scheduling` entirely immutable).
- Lines 607–610 state: *"If a downscale would make the represented pod count smaller than an explicitly configured `Gang.minCount`, reconciliation fails validation... the user must lower that minimum before scaling below it."*
If `spec.scheduling` is immutable, the user will be blocked from lowering `Gang.minCount`, permanently wedging the JobSet during a downscale.
**Recommendation:**
1. Explicitly document that for `ElasticJobSets`, `minCount` should be omitted so it automatically defaults and adjusts to `replicas × parallelism`.
2. Alternatively, scope the CEL rule so that constraints and policies are immutable, but `minCount` can be adjusted.

Author: Good point. I think for now, we could just require elastic jobs not set MinCount. 
That is what my implementation expects. 

We can always relax this over time but not clear on priority.

- **Grouped ReplicatedJobs** (a `replicatedJobPolicies` entry's `targetReplicatedJob` names more
  than one ReplicatedJob): the target names joined with `-`, e.g. `<target-1>-<target-2>`.
- **Per-Job mode / Gang-of-Gangs** (a `replicatedJobPolicies` entry sets `jobSchedulingPolicy`):
  one item per replica, `<replicatedjob-name>-<job-index>`.

Generated WorkloadItem names must be unique within a Workload — for example, a ReplicatedJob
literally named `leader-worker` could collide with a group formed from `leader` and `worker`.
The controller detects this and rejects reconciliation with an error naming the collision,
rather than silently merging or overwriting a PodGroupTemplate.

The PodGroup object created for each WorkloadItem follows the same split:

- For the top-level item (composite Gang policy, no sequenced startup), the PodGroup reuses the
  JobSet name directly: `<jobset-name>`.
- For every other item, the PodGroup is named `<jobset-name>-<item-name>`. If that exceeds the
  63-character DNS label limit, the JobSet name is truncated and a 10-character SHA-1 hash of
  `<jobset-name>/<item-name>` is appended instead, so the name stays deterministic and unique:

>   `<truncated-jobset-name>-<hash>`.

Reviewer: @kannon92 Shall we also say how we are going to cut the ReplicatedJob names to define PodGroup name as we discussed today?
Also, what if users have JobSet with 63 ReplicatedJobs that they want to group under a single PodGroup?

Author: I think the above explains this well.

> For every other item, the PodGroup is named `<jobset-name>-<item-name>`. If that exceeds the
  63-character DNS label limit, the JobSet name is truncated and a 10-character SHA-1 hash of
  `<jobset-name>/<item-name>` is appended instead, so the name stays deterministic and unique:
  `<truncated-jobset-name>-<hash>`.

Reviewer: Does this mean that if the PodGroup name exceeds the limit, we won’t include the ReplicatedJob name in it?
If so, to understand the relationship between a PodGroup and its ReplicatedJobs, would users need to look at the created Workload object?

Author: Yea I mean this is similar pattern with pods of job. If I create a name with really long job name the pods could truncate the name. 

You can probably use a selector to get all podgroups that correspond to a replicatedJob.

We may have to check but someone could find it in the namespace.

Author: For job, you can get the podgroups that correspond to the job via something like this:

```bash
kubectl get podgroups -n default -o json | jq -r '.items[] | select(.metadata.ownerReferences[]?.name=="gang-job") | .metadata.name'
```

I can add something like this for future extensions. I don't expect us to solve all UX issues in the first round of this.

Reviewer: Sounds good, let's collect users feedback to see if they really want additional informations (e.g. labels) to understand the relationship between CPG/PG and ReplicatedJobs.

   own WorkloadItem and PodGroup instead of sharing one PodGroup across the whole ReplicatedJob,
   so replicas can be gang-scheduled and preempted independently and can carry per-replica
   resource claims. `job` requires `targetReplicatedJobs` to name exactly one
   ReplicatedJob, and is mutually exclusive with the leaf-level `schedulingPolicy`,
   `schedulingConstraints`, `disruptionMode`, and `resourceClaims` fields on the same entry.
4. Each WorkloadItem is compiled with `workloadbuilder`. Because the current builder supports
   only single-item trees, per-ReplicatedJob and per-Job templates are merged into one Workload.
5. The controller creates one PodGroup for each resulting template and sets the JobSet as the
   controller owner of the Workload and PodGroups.
6. Child Jobs receive:
   - `scheduling.k8s.io/group-template-name`, identifying the template; and
   - a pod-template `schedulingGroup.podGroupName` pointing to the PodGroup.

The alpha implementation does not create a `CompositePodGroup`. It therefore does not set
`scheduling.k8s.io/parent-composite-podgroup`; that annotation refers to a
`CompositePodGroup`, not to the owning JobSet.

#### Naming Convention

> The Workload is always named after the JobSet (`<jobset-name>`). Because the current

Reviewer: @kannon92 I quickly review this again, we might want to as hash postfix to the Workload name.
What if users want to create any other object with the same name as JobSet (e.g. Job or Deployment).

```suggestion
The Workload is always named after the JobSet (`<jobset-name>-<hash>`). Because the current
```

Author: updated.

`workloadbuilder` only supports single-item builder trees, every WorkloadItem the JobSet
produces is built independently and then merged into that one Workload.

Each WorkloadItem — which compiles to one PodGroupTemplate — is named according to how it was
produced:

- **Top-level mode** (composite Gang or Basic policy, no sequenced startup, no targeted
  overrides): the JobSet name, `<jobset-name>`.
- **Per-ReplicatedJob mode** (a ReplicatedJob with no `replicatedJobs` entry targeting

>   it): the ReplicatedJob's own name, `<replicatedjob-name>`.

Reviewer: Should be with the postfix hash too, right?

Author: yea, updated.

// PodGroupTemplate/PodGroup per Job (replica) of the targeted ReplicatedJob, sized
// to that Job's own parallelism, rather than one PodGroup shared across all of the
// ReplicatedJob's replicas.
type JobScheduling struct {
    // +optional
    SchedulingPolicy *schedulingv1alpha3.PodGroupSchedulingPolicy `json:"schedulingPolicy,omitempty"`
    // +optional
    SchedulingConstraints *schedulingv1alpha3.PodGroupSchedulingConstraints `json:"schedulingConstraints,omitempty"`
    // +optional
    DisruptionMode *schedulingv1alpha3.DisruptionMode `json:"disruptionMode,omitempty"`
    // +optional
    // +listType=atomic
    // +kubebuilder:validation:MaxItems=4
    ResourceClaims []schedulingv1alpha3.PodGroupResourceClaim `json:"resourceClaims,omitempty"`
}
```

`JobSetSpec.Scheduling` is optional and immutable (enforced via a CEL `XValidation` rule rather
than webhook-only validation, per JobSet's API stability conventions).

### Defaulting

Defaulting applies only when `spec.scheduling` is present:

- Top-level Gang `minCount` defaults to the sum of `parallelism × replicas` across all
  ReplicatedJobs.
- Per-ReplicatedJob Gang `minCount` defaults to `parallelism × replicas` for that

>   ReplicatedJob.

Reviewer 3: Because `JobSetSpec.Scheduling` is strictly immutable via CEL (`self == oldSelf`), we should explicitly clarify here that `Gang.minCount` defaulting is performed **dynamically by the JobSet controller during compilation**, rather than by the mutating webhook on the `JobSet` object itself.
If the mutating webhook writes a defaulted `minCount` into `JobSet.spec.scheduling`, `minCount` becomes permanently materialized in `spec`. Later, if an ElasticJobSet scales down, the downscale is blocked because `minCount` can never be lowered in `spec`. Keeping `minCount` as `nil` in `JobSet.spec.scheduling` allows the controller to dynamically patch `PodGroup.spec.minCount` as parallelism scales.

Author: Good catch — updated the Defaulting section to make this explicit. Gang `minCount` is now documented as computed dynamically by the JobSet controller during compilation (written to `PodGroup.spec.minCount`), not materialized into the immutable `JobSet.spec.scheduling` by the mutating webhook, so ElasticJobSet scale-down stays possible. (AI tools assisted with this edit.)

//     replicatedJobs entry, since there is no top-level policy for an
//     untargeted ReplicatedJob to fall back to.
type JobSetScheduling struct {
    // schedulingPolicy defines the composite-level (level 1) scheduling policy for
    // the entire JobSet. Defaults to Gang when spec.scheduling is set but
    // schedulingPolicy is nil and replicatedJobs is not set.
    // Mutually exclusive with replicatedJobs: see the type-level comment.
    // +optional
    SchedulingPolicy *schedulingv1alpha3.PodGroupSchedulingPolicy `json:"schedulingPolicy,omitempty"`

    // schedulingConstraints defines composite-level (level 1) topology constraints
    // for the entire JobSet.
    // Mutually exclusive with replicatedJobs: see the type-level comment.
    // +optional
    SchedulingConstraints *schedulingv1alpha3.PodGroupSchedulingConstraints `json:"schedulingConstraints,omitempty"`

    // disruptionMode defines how the entire composite group (level 1) can be disrupted.
    // Mutually exclusive with replicatedJobs: see the type-level comment.
    // +optional
    DisruptionMode *schedulingv1alpha3.DisruptionMode `json:"disruptionMode,omitempty"`

> 

Reviewer: Let's ensure to add ResourceClaims to the level 1 too as @robert-bell suggested here: https://github.com/kubeflow/trainer/pull/3219#discussion_r3829614475

Author: done.

with JobSet.
An optional `spec.scheduling` field lets users configure gang scheduling, topology
constraints, disruption modes, and ReplicatedJob-level resource claims.

[kep-6089]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/6089-was-controller-apis

## Motivation

Distributed training and HPC workloads often need all pods admitted together, placed within a
specific topology domain, or disrupted as a group. JobSet currently has no native API for these
requirements.
WAS supplies standard scheduling APIs and a `workloadbuilder` library that JobSet
can compile into `Workload` and `PodGroup` resources.

### Goals

- Add optional, centralized scheduling configuration to `JobSetSpec`.
- Support Basic and Gang policies, topology constraints, disruption modes, and shared
  ReplicatedJob-level resource claims.
- Make the scheduling configuration available to integrations such as Kueue, so they can
  select the appropriate workload representation and admission behavior.
- Adopt the workload builder library for Workload/PodGroup creation and validation.

### Future Goals

- Creating a composite/parent PodGroup hierarchy in the alpha implementation.

> Moving this to a future goal because Composite PodGroup is alpha in 1.37.

Reviewer 4: But at the same time, we want to design a future-proof API that will be backward compatible with CompositePodGroup, right?

Author: Sounds like we have that.

Author: Future goals means its part of this KEP and we have a sketch of the API.

`targetReplicatedJobs`. When enabled, the JobSet controller builds one `Workload`, materializes
one `PodGroup` per scheduling template, and maps child Jobs and Pods to those PodGroups.

### User Stories

| User need | Behavior |
|---|---|
| Gang-schedule a complete training JobSet | A top-level Gang policy creates one PodGroup whose `minCount` is the total pod count. |
| Schedule groups independently | A `replicatedJobs` entry creates one PodGroup per targeted ReplicatedJob; Basic, Gang, topology, disruption, and resource-claim settings can be overridden. |
| Use an independent driver and constrained workers | The driver can use Basic scheduling while workers use a topology-constrained Gang PodGroup. These groups are independent in alpha; there is no composite PodGroup. |
| Keep existing JobSets unchanged | A JobSet without `spec.scheduling` creates no WAS resources. |
| Avoid reserving resources for suspended workloads | Suspended JobSets have their Workload and PodGroups deleted; they are recreated on resume. |
| Combine sequencing with Gang scheduling | `DependsOn` and `InOrder` startup use one PodGroup per ReplicatedJob to avoid deadlock. |

> | Scale an elastic workload | Changes to the represented pod count update the generated Gang `minCount`; other template changes recreate the owned scheduling resources. |

Reviewer 4: > other template changes recreate the owned scheduling resources.

I'm not sure I understand this. Recreating scheduling resources (PodGroups) would require recreating the pods—is that what we want to do here?

Author: I think this should say that other template changes require recreating the jobset.

Author: updated.

> | Scale an elastic workload | Changes to the represented pod count update the generated Gang `minCount`; other template changes recreate the owned scheduling resources. |

Author: ```suggestion
| Scale an elastic workload | Changes to the represented pod count update the generated Gang `minCount`; other template changes require deleting and recreating the jobset with updated values |
```

At a high level, the API looks like:

```yaml
scheduling:
  schedulingPolicy: .. (if gang, all pods scheduled together)
  schedulingConstraints: ... (topology for all jobs within JobSet)
  disruptionMode: ... (disruption for all pods within jobset)
  replicatedJobs:
    targetReplicatedJobs: ..(the below policies apply to these group of replicated jobs)
    schedulingPolicy: .. (gang within all replicas of the ReplicatedJob)
    schedulingConstraints: .. (tas requirements for all replicas of ReplicatedJob)
    disruptionMode: .. (disruption within replicas of replicated job)
    job:
      gang requirements at the job level
```

With the `JobSetGangOfGangs` feature gate enabled and the CompositePodGroups (CPG) feature,
the top level scheduling field corresponds to the root level.

With compositePodGroups, the top level policy would be a CPG object.
The group of replicated jobs within a `replicatedJobs` entry would correspond to a CPG also.

And then at the job level, these would be separate PodGroups.

> **Note (future work, Alpha 2):** JobSet's `replicatedJobs` intentionally allows
> heterogeneous ReplicatedJobs (for example a CPU driver alongside topology-constrained

> > accelerator workers) to be configured under one JobSet. Some downstream topology-aware

Reviewer 4: > Some downstream topology-aware
> scheduling plugins assume the child `PodGroups` under a `CompositePodGroup` are structurally
> homogeneous in order to compute placements efficiently.

Is this actually the case? We don't do anything like this in kube-scheduler today. While it's a valid optimization we could explore in the future, the scheduler shouldn't assume a specific API structure. Instead, it should dynamically detect homogeneity (e.g., that all CPG children are identical) and apply a more optimal algorithm when applicable. If our intent was to strictly enforce that child PodGroups under a CompositePodGroup must be structurally homogeneous, we would have designed the CPG/PG API differently.

Author: I'm actually not sure what @jianqiaol meant by downstream scheduling plugins. It could be Kueue's TAS or internal scheduling tools. 

I don't think he meant kubescheduler.

Author: @jianqiaol can you clarify what you meant here?

Reviewer 3: Thanks for clarifying @mm4tt, and sorry for the confusion @kannon92!

This was a drafting placeholder for Alpha 2. The note was inspired by early accelerator/TPU placement discussions that assumed identical slice shapes for solver heuristics.

I agree with @mm4tt: upstream `CompositePodGroup` supports heterogeneous children, and any homogeneity optimizations can be detected dynamically by scheduler plugins. 
Let's clean up this placeholder to simply state that multi-level `CompositePodGroup` compilation details will be finalized in Alpha 2.

Author: updated.

    SchedulingConstraints *schedulingv1alpha3.PodGroupSchedulingConstraints `json:"schedulingConstraints,omitempty"`
    // +optional
    DisruptionMode *schedulingv1alpha3.DisruptionMode `json:"disruptionMode,omitempty"`
    // +optional
    // +listType=atomic
    // +kubebuilder:validation:MaxItems=4
    ResourceClaims []schedulingv1alpha3.PodGroupResourceClaim `json:"resourceClaims,omitempty"`
}
```

`JobSetSpec.Scheduling` is optional and immutable (enforced via a CEL `XValidation` rule rather
than webhook-only validation, per JobSet's API stability conventions).

### Defaulting

Defaulting applies only when `spec.scheduling` is present:

- Top-level Gang `minCount` defaults to the sum of `parallelism × replicas` across all
  ReplicatedJobs.
- Per-ReplicatedJob Gang `minCount` defaults to `parallelism × replicas` for that
  ReplicatedJob.

Gang `minCount` defaulting is performed **dynamically by the JobSet controller during
compilation** — the controller computes the represented pod count and writes it to
`PodGroup.spec.minCount` — rather than by the mutating webhook writing a value back into
`JobSet.spec.scheduling`. Because `JobSetSpec.Scheduling` is strictly immutable

> (CEL `self == oldSelf`), a webhook-materialized `minCount` in `spec` could never be lowered

Reviewer 4: >  Because `JobSetSpec.Scheduling` is strictly immutable (CEL `self == oldSelf`),

How will this work with suspend? I believe that once the job is suspended, we should be able to completely change the scheduling configuration.

Author: Good point. We may need to remove the immutable tag and enforce via webhook.

Forbid changes once suspend is false.
Allow modifications once it is suspended.

Reviewer 4: Just to make it clear - I think it's completely fine starting small and keeping it immutable in alpha, with a plan to support modifications during suspend in future releases.

Author: I updated the KEP to reflect this. I'd like suspend supporting from the beginning.

   WorkloadItem represents the entire JobSet. For a Gang policy, its PodGroup uses the total
   represented pod count as `minCount`; for a Basic policy, the PodGroup has no `minCount` and
   each pod is admitted independently.
2. **Per-ReplicatedJob mode** is selected when overrides exist or startup is sequenced. One
   WorkloadItem is built for each ReplicatedJob (or, when a `replicatedJobs` entry targets
   more than one ReplicatedJob, one shared WorkloadItem for that group). A targeted policy takes
   precedence over composite defaults.
3. **Per-Job mode (Gang-of-Gangs)** is selected per `replicatedJobs` entry when that
   entry sets `job`. Each replica (Job) of the targeted ReplicatedJob gets its
   own WorkloadItem and PodGroup instead of sharing one PodGroup across the whole ReplicatedJob,
   so replicas can be gang-scheduled and preempted independently and can carry per-replica
   resource claims. `job` requires `targetReplicatedJobs` to name exactly one
   ReplicatedJob, and is mutually exclusive with the leaf-level `schedulingPolicy`,
   `schedulingConstraints`, `disruptionMode`, and `resourceClaims` fields on the same entry.

> 4. Each WorkloadItem is compiled with `workloadbuilder`. Because the current builder supports

Reviewer 4: @helayoty  - we should add a support for that. Could you open an issue?
